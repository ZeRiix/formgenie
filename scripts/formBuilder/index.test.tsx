import { useFormBuilder, type GetRef } from "./index";
import { useInput } from "./useInput";
import { render, waitFor, fireEvent } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import { createRef } from "react";
import { z } from "zod";

function InputTest(
	props: {
		label: string;
		value: string;
		onChange(value: any): void;
		placeholder: string;
		tt: string;
	},
) {
	return (
		<>
			<input
				onChange={(event) => void props.onChange(event.target.value)}
				value={props.value}
				placeholder={props.placeholder}
				name={props.label}
			/>
			<p>{props.tt}</p>
		</>
	);
}

describe("useBuilderForm", () => {
	it("should return the correct template", () => {
		const Form = useFormBuilder({
			name: useInput(
				InputTest,
				{
					defaultValue: "",
					label: "Name",
					props: {
						placeholder: "Type your name",
						tt: "cc les amis",
					},
				},
			),
			age: useInput(
				(props) => (
					<>
						<input
							onChange={(event) => void props.onChange(event.target.value)}
							value={props.value}
							placeholder={props.placeholder}
							name={props.name}
						/>
					</>
				),
				{
					label: "Age",
					defaultValue: 0,
					props: {
						placeholder: "Type your age",
					},
				},
			),
		});

		const componentRender = render(
			<Form ref={null} />,
		);

		expect(componentRender.asFragment()).toMatchSnapshot();
	});

	it("test default values", () => {
		const FormTest = useFormBuilder({
			name: useInput(
				InputTest,
				{
					defaultValue: "test",
					label: "Name",
					props: {
						placeholder: "Type your name",
						tt: "cc les amis",
					},
				},
			),
		});

		const formRef = createRef<GetRef<typeof FormTest>>();

		render(
			<FormTest ref={formRef} />,
		);

		expect(formRef.current?.values.name).toMatchSnapshot();
	});

	it("test check error", async() => {
		const MIN_SCHEMA = 5;
		const FormTest = useFormBuilder({
			name: useInput(
				InputTest,
				{
					defaultValue: "test",
					label: "Name",
					props: {
						placeholder: "Type your name",
						tt: "cc les amis",
					},
					zodSchema: z.string().min(MIN_SCHEMA),
				},
			),
		});

		const formRef = createRef<GetRef<typeof FormTest>>();

		render(
			<FormTest ref={formRef} />,
		);

		expect(await formRef.current?.check()).toMatchSnapshot();
	});

	it("test check good", async() => {
		const MIN_SCHEMA = 3;
		const FormTest = useFormBuilder({
			name: useInput(
				InputTest,
				{
					defaultValue: "test",
					label: "Name",
					props: {
						placeholder: "Type your name",
						tt: "cc les amis",
					},
					zodSchema: z.string().min(MIN_SCHEMA),
				},
			),
		});

		const formRef = createRef<GetRef<typeof FormTest>>();

		render(
			<FormTest ref={formRef} />,
		);

		expect(await formRef.current?.check()).toMatchSnapshot();
	});

	it("test check good without zodSchema", async() => {
		const FormTest = useFormBuilder({
			name: useInput(
				InputTest,
				{
					defaultValue: "test",
					label: "Name",
					props: {
						placeholder: "Type your name",
						tt: "cc les amis",
					},
				},
			),
		});

		const formRef = createRef<GetRef<typeof FormTest>>();

		render(
			<FormTest ref={formRef} />,
		);

		expect(await formRef.current?.check()).toMatchSnapshot();
	});

	it("test reset", async() => {
		const FormTest = useFormBuilder({
			name: useInput(
				InputTest,
				{
					defaultValue: "test",
					label: "Name",
					props: {
						placeholder: "Type your name",
						tt: "cc les amis",
					},
				},
			),
		});

		const formRef = createRef<GetRef<typeof FormTest>>();

		const { getByPlaceholderText } = render(
			<FormTest ref={formRef} />,
		);

		const input = getByPlaceholderText("Type your name") as HTMLInputElement;

		expect(input.value).toMatchSnapshot();

		fireEvent.change(input, { target: { value: "new value" } });
		expect(input.value).toMatchSnapshot();

		formRef.current?.reset();

		await waitFor(() => {
			expect(input.value).toMatchSnapshot();
		});
	});
});
