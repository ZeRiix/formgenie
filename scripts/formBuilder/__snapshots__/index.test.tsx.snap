// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useBuilderForm > should return the correct template 1`] = `
<DocumentFragment>
  <form>
    <div>
      <label
        for="Name"
      >
        Name
      </label>
      <input
        name="Name"
        placeholder="Type your name"
        value=""
      />
      <p>
        cc les amis
      </p>
    </div>
    <div>
      <label
        for="Age"
      >
        Age
      </label>
      <input
        placeholder="Type your age"
        value="0"
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`useBuilderForm > test check error 1`] = `
{
  "name": undefined,
}
`;

exports[`useBuilderForm > test check good 1`] = `
{
  "name": "test",
}
`;

exports[`useBuilderForm > test check good without zodSchema 1`] = `
{
  "name": "test",
}
`;

exports[`useBuilderForm > test default values 1`] = `"test"`;

exports[`useBuilderForm > test edit values 1`] = `"test"`;

exports[`useBuilderForm > test edit values 2`] = `"test"`;

exports[`useBuilderForm > test reset 1`] = `"test"`;

exports[`useBuilderForm > test reset 2`] = `"new value"`;

exports[`useBuilderForm > test reset 3`] = `"new value"`;
